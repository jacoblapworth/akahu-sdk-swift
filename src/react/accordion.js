import XUIAccordion from './components/accordion/XUIAccordion';

export default XUIAccordion;
