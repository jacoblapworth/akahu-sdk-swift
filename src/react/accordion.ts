import XUIAccordion from './components/accordion/XUIAccordion';
import XUIAccordionItem from './components/accordion/XUIAccordionItem';

export { XUIAccordion as default, XUIAccordion, XUIAccordionItem };
