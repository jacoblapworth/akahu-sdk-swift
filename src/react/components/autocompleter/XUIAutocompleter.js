import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import throttle from 'lodash.throttle';
import Picklist from '../picklist/Picklist';
import XUILoader from '../loader/XUILoader';
import DropDown from '../dropdown/DropDown';
import DropDownToggled from '../dropdown/DropDownToggled';
import XUITextInput from '../textInput/XUITextInput';
import { ns } from '../helpers/xuiClassNamespace';

/**
 * Keyboard bindings to ignore. Space doesn't select in an autocompleter; left and
 * right arrow keys should move cursor in the input
 *
 * @private
 * @type {Array}
 */
const ignoreKeyboardEvents = [32, 37, 39];

export default class XUIAutocompleter extends PureComponent {
	constructor(props) {
		super(props);
		this.state = {
			focused: false,
			value: props.searchValue,
		};
		this.bindOnChange(props.searchThrottleInterval);
	}

	componentDidMount() {
		this.calculatePlaceholderWidth();
	}

	componentDidUpdate(prevProps) {
		const {
			pills,
			disableWrapPills,
			searchThrottleInterval,
			searchValue,
			placeholder,
		} = this.props;
		if (prevProps.searchThrottleInterval !== searchThrottleInterval) {
			this.bindOnChange(searchThrottleInterval);
		}
		if (prevProps.searchValue !== searchValue) {
			// TODO: Investigate whether setState can be moved
			this.setState({ // eslint-disable-line
				value: searchValue,
			});
		}
		if (prevProps.placeholder !== placeholder) {
			this.calculatePlaceholderWidth();
		}
		const morePillsExist = React.Children.count(pills) > React.Children.count(prevProps.pills);
		if (morePillsExist && disableWrapPills) {
			this.noWrapPillContainer.scrollLeft = this.noWrapPillContainer.scrollWidth;
		}
	}

	/**
	 * Bind an optionally throttled onSearch handler to the component instance.
	 *
	 * @private
	 * @param {number} interval
	 */
	bindOnChange = interval => {
		const { onSearch } = this.props;
		if (onSearch) {
			const throttled = interval
				? throttle(onSearch, interval, { trailing: true })
				: onSearch;
			this.throttledOnChange = event => {
				event.persist();
				this.setState({
					value: event.target.value,
				});
				throttled(event.target.value);
			};
		} else {
			this.throttledOnChange = undefined;
		}
	};

	calculatePlaceholderWidth = () => {
		if (this.placeholder != null) {
			const placeholderWidth = getComputedStyle(this.placeholder).width;
			const inputStyle = getComputedStyle(this.inputNode);
			const inputWidth = `${
				parseFloat(inputStyle.paddingLeft)
				+ parseFloat(inputStyle.paddingRight)
				+ parseFloat(placeholderWidth)
			}px`;
			if (this.state.inputWidth !== inputWidth) {
				this.setState({
					inputWidth,
				});
			}
		}
	};

	/**
	 * @public
	 * Set the state as not hidden in order to toggle the list open.
	 */
	openDropDown = () => {
		this.ddt.openDropDown();
	};

	/**
	 * @public
	 * Set the state as hidden in order to toggle the list closed.
	 */
	closeDropDown = () => {
		this.ddt.closeDropDown();
	};

	/**
	 * @public
	 * Manually highlight an item in the list for selection.
	 */
	highlightItem = item => {
		this.dropdown.highlightItem(item);
	};

	/**
	 * @public
	 * Focuses the text input
	 */
	focusInput = () => {
		this.inputNode.focus();
	};

	/**
	 * @public
	 * If a onHighlightChange prop is passed to the completer, it's called passing
	 * in the highlighted item.
	 *
	 * @param {item} Object
	 */
	onHighlightChange = item => {
		this.props.onHighlightChange && this.props.onHighlightChange(item);
	};

	onInputKeyDown = event => {
		const {
			onBackspacePill,
			pills,
		} = this.props;
		if (this.ddt.isDropDownOpen()) {
			this.dropdown.onKeyDown(event);
		}

		if (
			event.key === 'Backspace' &&
			this.inputNode.value === '' &&
			onBackspacePill &&
			pills &&
			(pills.length > 0 || React.isValidElement(pills))
		) {
			onBackspacePill();
		}
	};

	onInputFocus = () => {
		if (!this.state.focused) {
			this.openDropDown();
		}
	};

	onFocus = () => {
		this.focusInput();
		this.setState({
			focused: true,
		});
	};

	onBlur = () => {
		setTimeout(() => {
			if (this.rootNode && !this.rootNode.contains(document.activeElement)) {
				this.setState({
					focused: false,
				});
			}
		}, 333);
	};

	renderPills = () => {
		const {
			disableWrapPills,
			pills,
		} = this.props;
		return disableWrapPills ? (
			<div
				className={`${ns}-autocompleter--pills-nopillwrap`}
				ref={nwpc => this.noWrapPillContainer = nwpc}
			>
				{pills}
			</div>
		) : pills;
	}

	render() {
		const completer = this;
		const {
			qaHook,
			pills,
			leftElement,
			inputContainerClassName,
			disableWrapPills,
			inputClassName,
			openOnFocus,
			triggerClassName,
			placeholder,
			rightElement,
			isDisabled,
			inputLabelText,
			isInputLabelHidden,
			inputProps,
			maxLength,
			inputId,
			dropdownId,
			onOptionSelect,
			dropdownClassName,
			dropdownSize,
			dropdownFixedWidth,
			footer,
			loading,
			children,
			className,
			id,
			onOpen,
			onClose,
			closeOnTab,
			closeOnSelect,
			forceDesktop,
			matchTriggerWidth,
		} = this.props;
		const { value, focused, inputWidth } = this.state;
		let inputQaHook = null;
		let listQaHook = null;
		let containerQaHook = null;
		let dropdownQaHook = null;
		if (qaHook) {
			inputQaHook = `${qaHook}`; // TODO: Investigate whether we should add --input here in 14
			listQaHook = `${qaHook}--list`;
			containerQaHook = `${qaHook}--container`;
			dropdownQaHook = `${qaHook}--dropdown`;
		}
		const hasPills = pills != null && (pills.length > 0 || React.isValidElement(pills));
		const textInputLeftElement = hasPills ? this.renderPills() : leftElement;

		const containerClassNames = cn(
			inputContainerClassName,
			{
				[`${ns}-autocompleter--trigger-pillwrap`]: hasPills && !disableWrapPills,
			},
		);

		const inputClassNames = cn(
			inputClassName,
			`${ns}-autocompleter--textinput`,
			{ [`${ns}-padding-left-small`]: hasPills },
		);

		const trigger = (
			<div
				ref={tg => this.trigger = tg}
				onFocus={openOnFocus ? this.onInputFocus : null}
				className={triggerClassName}
			>
				<div
					ref={p => this.placeholder = p}
					className={`${ns}-autocompleter--textinputplaceholder`}
					aria-hidden
				>
					{placeholder}
				</div>
				<XUITextInput
					leftElement={textInputLeftElement}
					rightElement={rightElement}
					containerClassName={containerClassNames}
					inputClassName={inputClassNames}
					inputRef={i => this.inputNode = i}
					placeholder={placeholder}
					value={value || ''}
					onChange={this.throttledOnChange}
					onKeyDown={this.onInputKeyDown}
					qaHook={inputQaHook}
					isDisabled={isDisabled}
					labelText={inputLabelText}
					isLabelHidden={isInputLabelHidden}
					inputProps={{
						...inputProps,
						'maxLength': maxLength,
						'id': inputId,
						'role': 'textbox',
						'aria-multiline': false,
						'aria-autocomplete': 'list',
						'style': {
							flexBasis: inputWidth,
						},
					}}
				/>
			</div>
		);

		const dropdown = (
			<DropDown
				ref={c => completer.dropdown = c}
				ignoreKeyboardEvents={ignoreKeyboardEvents}
				id={dropdownId}
				onSelect={onOptionSelect}
				hasKeyboardEvents={false}
				className={dropdownClassName}
				qaHook={listQaHook}
				restrictFocus={false}
				size={dropdownSize}
				fixedWidth={dropdownFixedWidth}
				footer={footer}
				onHighlightChange={completer.onHighlightChange}
			>
				{loading ? <Picklist><XUILoader /></Picklist> : children}
			</DropDown>
		);

		const classNames = cn(
			className,
			focused && `${ns}-autocompleter--trigger-focus`,
		);

		return (
			<div
				ref={c => completer.rootNode = c}
				className={classNames}
				onFocus={this.onFocus}
				onBlur={this.onBlur}
				data-automationid={containerQaHook}
				id={id}
			>
				<DropDownToggled
					ref={c => completer.ddt = c}
					trigger={trigger}
					dropdown={dropdown}
					onOpen={onOpen}
					onClose={onClose}
					closeOnTab={closeOnTab}
					closeOnSelect={closeOnSelect}
					triggerClickAction="none"
					forceDesktop={forceDesktop}
					matchTriggerWidth={matchTriggerWidth && !dropdownSize}
					qaHook={dropdownQaHook}
					isBlock
					ariaRole="combobox"
				/>
			</div>
		);
	}
}

// TODO: Rename `pills` to `leftElement` for XUI 14.
XUIAutocompleter.propTypes = {
	/** Callback to handle when an option has been selected from the dropdown */
	onOptionSelect: PropTypes.func,

	/** Callback to handle when a pill has been backspaced */
	onBackspacePill: PropTypes.func,

	/** When set to true a loader will be displayed instead of the picklist items.
	 * State for this should be managed externally and it's defaulted to false.
	 */
	loading: PropTypes.bool,

	/** ID to be added to the root node of the completer */
	id: PropTypes.string,

	/** ID to be added to the dropdown element of the completer */
	dropdownId: PropTypes.string,

	/** Value that should be inside the input. */
	searchValue: PropTypes.string,

	/** CSS class(es) to go on the wrapping DOM node */
	className: PropTypes.string,

	/** CSS class(es) to go on the dropdown list */
	dropdownClassName: PropTypes.string,

	/** CSS class(es) to go on the input */
	inputClassName: PropTypes.string,

	/** CSS class(es) to go on the input container component */
	inputContainerClassName: PropTypes.string,

	/** Label to show above the input */
	inputLabelText: PropTypes.string,

	/** Should label be applied as an aria-label, rather than being visibly displayed. */
	isInputLabelHidden: PropTypes.bool,

	/** Attributes to set on the native input element */
	inputProps: PropTypes.object,

	/** CSS class(es) to go on the trigger element which contains the input and pills */
	triggerClassName: PropTypes.string,

	/** Placeholder for the input */
	placeholder: PropTypes.string,

	/** Max length of the input */
	maxLength: PropTypes.number,

	/** A set of pills to show next to input. Useful for showing what was selected in a multi-select.
	 * Can also be used similarly to `XUITextInput`'s `leftElement`. */
	pills: PropTypes.oneOfType([PropTypes.element, PropTypes.arrayOf(PropTypes.element)]),

	/** Right element to render within the `XUITextInput` component */
	rightElement: PropTypes.node,

	/** Left element to render within the `XUITextInput` component. Should not be used together with
	 * the `pills` prop */
	leftElement: PropTypes.node,

	/** Callback for when the list opens */
	onOpen: PropTypes.func,

	/** Callback for when the list closes */
	onClose: PropTypes.func,

	/** Callback for when the user types into the search box */
	onSearch: PropTypes.func,

	/** If you want to throttle the input's onChange handler, put the throttle interval here */
	searchThrottleInterval: PropTypes.number,

	/** Maps to the 'size' property of the dropdown component. */
	dropdownSize: PropTypes.oneOf(['small', 'medium', 'large', 'xlarge']),

	/** Maps to the `closeOnSelect` property of the DropDownToggled component. */
	closeOnSelect: PropTypes.bool,

	/** Maps to the `closeOnTab` property of the DropDownToggled component. Set to false, if you've
	 * supplied a footer element with any links or interaction. */
	closeOnTab: PropTypes.bool,

	/** When set to true the dropdown will automatically open when the input is given focus. */
	openOnFocus: PropTypes.bool,

	/** Force the desktop user experience, even if the viewport is narrow enough for mobile. */
	forceDesktop: PropTypes.bool,

	/** If a size is set, this will force the dropdown to that size instead of setting it as a
	 * max width. */
	dropdownFixedWidth: PropTypes.bool,

	/**
	 * Setting to false will allow the dropdown's width to be set independent of the trigger width.
	 * Note: Setting this to true will override any size prop on DropDown.  XUI design has also decided
	 * to keep a minimum width on the dropdown, so dropdown may not match the width of narrow triggers.
	 */
	matchTriggerWidth: PropTypes.bool,

	/** Whether the pills should wrap instead of scroll on overflow */
	disableWrapPills: PropTypes.bool,

	/** Whether to render as disabled */
	isDisabled: PropTypes.bool,

	/** ID to apply to the input element. Useful for labels. */
	inputId: PropTypes.string,

	/** A footer element can be added. */
	footer: PropTypes.element,

	/** Callback for when the highlighted item changes. */
	onHighlightChange: PropTypes.func,

	qaHook: PropTypes.string,
	children: PropTypes.node,
};

XUIAutocompleter.defaultProps = {
	loading: false,
	searchThrottleInterval: 0,
	openOnFocus: false,
	closeOnTab: true,
	forceDesktop: true,
	dropdownFixedWidth: false,
	matchTriggerWidth: true,
	disableWrapPills: false,
};
