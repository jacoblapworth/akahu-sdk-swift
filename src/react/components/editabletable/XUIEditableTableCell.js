import React from 'react';
import PropTypes from 'prop-types';

import cn from 'classnames';

import { tableName } from './private/constants';

const baseName = `${tableName}cell`;

const XUIEditableTableCell = ({ children, className, width, ...spreadProps }) => {
  return (
    <td className={cn(baseName, className)} style={{ width }} {...spreadProps}>
      {children}
    </td>
  );
};

XUIEditableTableCell.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  width: PropTypes.string,
};

export default XUIEditableTableCell;
