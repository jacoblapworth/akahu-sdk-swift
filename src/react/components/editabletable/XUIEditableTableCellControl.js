import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';

import { tableName } from './private/constants';
import PortalFocus from './private/PortalFocus';
import XUIEditableTableContext from './contexts/XUIEditableTableContext';
import EditableTableCellContext from '../../contexts/EditableTableCellContext';
import XUIEditableTableCell from './XUIEditableTableCell';
import { getAriaAttributes } from '../controlwrapper/XUIControlWrapper';
import generateIds, { generateIdsFromControlId } from '../controlwrapper/helpers';

const baseName = `${tableName}cell`;

const XUIEditableTableCellControl = ({
  children,
  className,
  cellIds,
  isDisabled,
  isFocused,
  isInvalid,
  validationMessage,
  ...spreadProps
}) => {
  const cellRef = React.useRef();
  const controlBaseName = `${baseName}-control`;
  const { tableRef } = React.useContext(XUIEditableTableContext);

  const showValidationMessage = isInvalid && validationMessage;
  const wrapperIds = cellIds?.control
    ? generateIdsFromControlId(cellIds?.control)
    : generateIds(cellIds?.wrapper);
  const cellAttributes =
    showValidationMessage && getAriaAttributes(wrapperIds, { isInvalid, validationMessage });

  return (
    <XUIEditableTableCell
      className={cn(
        className,
        controlBaseName,
        isDisabled && `${controlBaseName}-is-disabled`,
        isFocused && `${controlBaseName}-is-focused`,
        isInvalid && `${controlBaseName}-is-invalid`,
      )}
      ref={cellRef}
      {...spreadProps}
    >
      <EditableTableCellContext.Provider
        value={{
          cellAttributes,
          cellRef,
          useCellStyling: true,
        }}
      >
        <div className={`${baseName}--border`}>
          {children}
          {showValidationMessage && (
            <div
              className={cn(
                `${baseName}--validation`,
                isInvalid && `${baseName}--validation-is-invalid`,
              )}
              id={wrapperIds.message}
            >
              {validationMessage}
            </div>
          )}
          {isFocused && <PortalFocus cellRef={cellRef} isFocused={isFocused} tableRef={tableRef} />}
        </div>
      </EditableTableCellContext.Provider>
    </XUIEditableTableCell>
  );
};

XUIEditableTableCellControl.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  /**
   * Used to style the cell as disabled.
   */
  isDisabled: PropTypes.bool,
  /**
   * Used to style the cell as focused.
   */
  isFocused: PropTypes.bool,
  /**
   * Used to style the cell as invalid.
   */
  isInvalid: PropTypes.bool,
  /**
   * Validation message to show under the input if `isInvalid` is true.
   */
  validationMessage: PropTypes.string,
  /** Id of the control cell */
  cellIds: PropTypes.shape({
    wrapper: PropTypes.string,
    control: PropTypes.string,
  }),
};

export default XUIEditableTableCellControl;
