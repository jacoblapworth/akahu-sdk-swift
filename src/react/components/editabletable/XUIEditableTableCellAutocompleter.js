import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';

import XUIEditableTableCellControl from './XUIEditableTableCellControl';
import XUIAutocompleter from '../autocompleter/XUIAutocompleter';
import { tableVariantClassNames } from './private/constants';
import { fixedWidthDropdownSizes } from '../dropdown/private/constants';
import { ns } from '../helpers/xuiClassNamespace';

const baseName = `${tableVariantClassNames.editable}cellautocompleter`;

class XUIEditableTableCellAutocompleter extends Component {
  completerRef = React.createRef();

  state = {
    isFocused: false,
  };

  /**
   * Records the focus state onBlur, before calling any user-supplied handlers.
   * @param {event} event
   */
  composedOnBlur = event => {
    this.setState({ isFocused: false });
    this.props.inputProps && this.props.inputProps.onBlur && this.props.inputProps.onBlur(event);
  };

  /**
   * Records the focus state onFocus, before calling any user-supplied handlers.
   * @param {event} event
   */
  composedOnFocus = event => {
    this.setState({ isFocused: true });
    this.props.inputProps && this.props.inputProps.onFocus && this.props.inputProps.onFocus(event);
  };

  /**
   * @public
   * Set the state as not hidden in order to toggle the list open.
   */
  openDropdown = () => {
    this.completerRef.current.openDropdown();
  };

  /**
   * @public
   * Set the state as hidden in order to toggle the list closed.
   */
  closeDropdown = () => {
    this.completerRef.current.closeDropdown();
  };

  /**
   * Focuses the input inside the cell, before calling any user-supplied handlers.
   */
  composedOnClick = event => {
    const selection = getSelection();
    (selection.toString().length === 0 || !event.target?.contains(selection.focusNode)) &&
      this.focusInput();

    this.props.cellProps?.onClick?.(event);
  };

  /**
   * Focuses the input inside the cell, before calling any user-supplied handlers.
   */
  composedOnMouseDown = event => {
    if (event.target.classList.contains(`${ns}-editabletablecell--validation`)) {
      // Close the dropdown when interacting with validation
      this.closeDropdown();
    }
    if (event.target === event.currentTarget) {
      // Don't lose focus when interacting with the table cell
      event.preventDefault();

      // DDT closes its dropdown when the window receives an onMouseDown from outside the dropdown.
      // When the dropdown is closed the trigger gains focus, so we need to send the focus to this
      // cell's input after other dropdowns have closed.
      setTimeout(this.focusInput);
    }

    this.props.cellProps?.onMouseDown?.(event);
  };

  /**
   * @public
   * Focus the input inside the cell.
   */
  focusInput = () => {
    this.completerRef.current?.focusInput();
  };

  render() {
    const {
      cellProps = {},
      className,
      inputProps,
      isDisabled,
      isInvalid,
      forceStatefulPicklist,
      triggerClassName,
      useNewFocusBehaviour,
      validationMessage,
      ...spreadProps
    } = this.props;

    return (
      <XUIEditableTableCellControl
        {...cellProps}
        cellIds={{
          wrapper: spreadProps.id,
          control: inputProps?.id,
        }}
        className={cn(baseName, cellProps.className)}
        isDisabled={isDisabled}
        isFocused={this.state.isFocused}
        isInvalid={isInvalid}
        onClick={this.composedOnClick}
        onMouseDown={this.composedOnMouseDown}
        validationMessage={validationMessage}
      >
        <XUIAutocompleter
          {...spreadProps}
          _useCellStyling
          className={cn(`${baseName}--control`, className)}
          forceStatefulPicklist={forceStatefulPicklist}
          inputProps={{
            ...inputProps,
            onBlur: this.composedOnBlur,
            onFocus: this.composedOnFocus,
          }}
          isDisabled={isDisabled}
          isInputLabelHidden
          isInvalid={isInvalid}
          ref={this.completerRef}
          triggerClassName={cn(`${baseName}--trigger`, triggerClassName)}
          useNewFocusBehaviour={useNewFocusBehaviour}
        />
      </XUIEditableTableCellControl>
    );
  }
}

XUIEditableTableCellAutocompleter.propTypes = {
  cellProps: PropTypes.object,
  children: PropTypes.node,

  /** CSS class(es) to go on the wrapping DOM node */
  className: PropTypes.string,

  /** Maps to the `closeOnSelect` property of the DropdownToggled component. */
  closeOnSelect: PropTypes.bool,

  /** Maps to the `closeOnTab` property of the DropdownToggled component. Set to false, if you've
   * supplied a footer element with any links or interaction. */
  closeOnTab: PropTypes.bool,

  /** Whether the pills should wrap instead of scroll on overflow */
  disableWrapPills: PropTypes.bool,

  /** CSS class(es) to go on the dropdown list */
  dropdownClassName: PropTypes.string,

  /** If a size is set, this will force the dropdown to that size instead of setting it as a
   * max width. */
  dropdownHasFixedWidth: PropTypes.bool,

  /** ID to be added to the dropdown element of the completer */
  dropdownId: PropTypes.string,

  /** Maps to the `size` property of the dropdown component. */
  dropdownSize: PropTypes.oneOf(Object.keys(fixedWidthDropdownSizes)),

  /** A footer element can be added. */
  footer: PropTypes.element,

  /** Force the desktop user experience, even if the viewport is narrow enough for mobile. */
  forceDesktop: PropTypes.bool,

  /** Force wrapping `XUIDropdownPanel` children in a `XUIStatefulPicklist` */
  forceStatefulPicklist: PropTypes.bool,

  /** ID to be added to the root node of the completer */
  id: PropTypes.string,

  /** CSS class(es) to go on the input */
  inputClassName: PropTypes.string,

  /** CSS class(es) to go on the input container component */
  inputContainerClassName: PropTypes.string,

  /** ID to apply to the input element. Useful for labels. */
  inputId: PropTypes.string,

  /** Label to show above the input */
  inputLabel: PropTypes.node,

  /**
   * Attributes to set on the native input element. <br>
   * **Note:**
   * *It's not recommended to pass `autoFocus` to `inputProps` as it hijacks the focus on load to focus that specific element.*
   */
  inputProps: PropTypes.object,

  /** Whether to render as disabled */
  isDisabled: PropTypes.bool,

  /** Whether the current input value is invalid */
  isInvalid: PropTypes.bool,

  /** When set to true a loader will be displayed instead of the picklist items.
   * State for this should be managed externally and it's defaulted to false.
   */
  isLoading: PropTypes.bool,

  /** Left element to render within the `XUITextInput` component. Should not be used together with
   * the `pills` prop */
  leftElement: PropTypes.node,

  /**
   * Accessibility label for the `<XUILoader>`. This is required if the
   * `isLoading` prop is set to `true`.
   * <br />
   * Recommended English value: *Loading*
   */
  loadingAriaLabel: PropTypes.string,

  /**
   * Setting this to `true` makes the dropdown as wide as the trigger (defaults to `true`). <br>
   * **Note:** *Setting this to `true` will override any `size` prop on Dropdown.* <br>
   * Setting this to `false` will allow the dropdown's width to be set independent of the trigger width. <br>
   * Setting this to `'min'` will set the dropdown's `min-width` to be the trigger width. <br/>
   * XUI design has also decided to keep a minimum width on the dropdown,
   * so dropdown may not match the width of narrow triggers (setting this to `'min'` will not override this).
   */
  matchTriggerWidth: PropTypes.oneOf([true, false, 'min']),

  /** Max length of the input */
  maxLength: PropTypes.number,

  /** Callback to handle when a pill has been backspaced */
  onBackspacePill: PropTypes.func,

  /** Callback for when the list closes */
  onClose: PropTypes.func,

  /** Callback for when the highlighted item changes. */
  onHighlightChange: PropTypes.func,

  /** Callback for adding additional onKeyDown functionality */
  onKeyDown: PropTypes.func,

  /** Callback for when the list opens */
  onOpen: PropTypes.func,

  /** Callback to handle when an option has been selected from the dropdown */
  onOptionSelect: PropTypes.func,

  /** Callback for when the user types into the search box.
   * The first argument passed in is the search term value. */
  onSearch: PropTypes.func,

  /** When set to true the dropdown will automatically open when the input is given focus. */
  openOnFocus: PropTypes.bool,

  /** A set of pills to show next to input. Useful for showing what was selected in a multi-select.
   * Can also be used similarly to `XUITextInput`'s `leftElement`. */
  pills: PropTypes.oneOfType([PropTypes.element, PropTypes.arrayOf(PropTypes.element)]),

  /** Placeholder for the input */
  placeholder: PropTypes.string,

  qaHook: PropTypes.string,

  /** Right element to render within the `XUITextInput` component */
  rightElement: PropTypes.node,

  /** The debounce timeout before onSearch is called. Set to 0 to disable debouncing */
  searchDebounceTimeout: PropTypes.number,

  /** Value that should be inside the input. */
  searchValue: PropTypes.string,

  /** CSS class(es) to go on the trigger element which contains the input and pills */
  triggerClassName: PropTypes.string,

  /** Whether or not to use the new focus behaviour - which treats dropdown navigation
   * like a `combobox` role. Defaults to `false`.
   */
  useNewFocusBehaviour: PropTypes.bool,

  /** Validation message to show under the input if `isInvalid` is true */
  validationMessage: PropTypes.node,
};

export default XUIEditableTableCellAutocompleter;

XUIEditableTableCellAutocompleter.defaultProps = {
  forceStatefulPicklist: false,
  matchTriggerWidth: true,
};
