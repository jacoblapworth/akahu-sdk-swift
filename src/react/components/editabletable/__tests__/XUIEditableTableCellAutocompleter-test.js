import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Enzyme, { mount, shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import toJson from 'enzyme-to-json';
import { nanoid } from 'nanoid';
import { axe, toHaveNoViolations } from 'jest-axe';

import XUIAutocompleter from '../../autocompleter/XUIAutocompleter';
import XUIPill from '../../pill/XUIPill';
import XUIEditableTableCell from '../XUIEditableTableCell';
import XUIEditableTableCellControl from '../XUIEditableTableCellControl';
import XUIEditableTableCellAutocompleter from '../XUIEditableTableCellAutocompleter';

jest.mock('nanoid');
nanoid.mockImplementation(() => 'testGeneratedId');

Enzyme.configure({ adapter: new Adapter() });
expect.extend(toHaveNoViolations);

describe('<XUIEditableTableCellAutocompleter />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<XUIEditableTableCellAutocompleter onSearch={() => {}} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  // prettier-ignore
  it('should pass accessibility testing', async () => {
    const wrapper = mount(
      <table>
        <tbody>
          <tr><XUIEditableTableCellAutocompleter inputLabel="Autocompleter cell" onSearch={() => {}} /></tr>
        </tbody>
      </table>,
    );
    const results = await axe(wrapper.html());
    expect(results).toHaveNoViolations();
  });

  describe('focusing', () => {
    it('lets XUIEditableTableCellControl know when the input is focused', () => {
      // Arrange
      const wrapper = mount(
        <table>
          <tbody>
            <tr>
              <XUIEditableTableCellAutocompleter onSearch={() => {}} />
            </tr>
          </tbody>
        </table>,
      );

      // Act
      wrapper.find('input').simulate('focus');

      // Assert
      expect(wrapper.find(XUIEditableTableCellControl).prop('isFocused')).toBeTruthy();
    });

    it('lets XUIEditableTableCellControl know when the input is blurred', () => {
      // Arrange
      const wrapper = mount(
        <table>
          <tbody>
            <tr>
              <XUIEditableTableCellAutocompleter onSearch={() => {}} />
            </tr>
          </tbody>
        </table>,
      );

      // Act
      wrapper.find('input').simulate('focus');
      wrapper.find('input').simulate('blur');

      // Assert
      expect(wrapper.find(XUIEditableTableCellControl).prop('isFocused')).toBeFalsy();
    });

    it('lets XUIEditableTableCellControl know when the input is disabled', () => {
      // Arrange
      const wrapper = mount(
        <table>
          <tbody>
            <tr>
              <XUIEditableTableCellAutocompleter onSearch={() => {}} isDisabled />
            </tr>
          </tbody>
        </table>,
      );

      // Assert
      expect(wrapper.find(XUIEditableTableCellControl).prop('isDisabled')).toBeTruthy();
    });

    it('lets XUIEditableTableCellControl know when the input is invalid', () => {
      // Arrange
      const wrapper = mount(
        <table>
          <tbody>
            <tr>
              <XUIEditableTableCellAutocompleter onSearch={() => {}} isInvalid />
            </tr>
          </tbody>
        </table>,
      );

      // Assert
      expect(wrapper.find(XUIEditableTableCellControl).prop('isInvalid')).toBeTruthy();
    });

    it('lets XUIEditableTableCellControl know when the input has a validation message', () => {
      // Arrange
      const expectedMessage = 'Test validation message';
      const wrapper = mount(
        <table>
          <tbody>
            <tr>
              <XUIEditableTableCellAutocompleter
                onSearch={() => {}}
                validationMessage={expectedMessage}
              />
            </tr>
          </tbody>
        </table>,
      );

      // Assert
      expect(wrapper.find(XUIEditableTableCellControl).prop('validationMessage')).toBe(
        expectedMessage,
      );
    });
  });

  it('spreads cellProps onto the table cell', () => {
    const wrapper = mount(
      <table>
        <tbody>
          <tr>
            <XUIEditableTableCellAutocompleter cellProps={{ width: '100px' }} onSearch={() => {}} />
          </tr>
        </tbody>
      </table>,
    );
    expect(wrapper.find(XUIEditableTableCell).getDOMNode().getAttribute('width')).toBe('100px');
  });

  it('spreads the rest of the props onto XUIEditableTableCellAutocompleter', () => {
    const spreadProps = {
      id: 'testId',
      onSearch: () => {},
      isDisabled: true,
    };

    const wrapper = mount(
      <table>
        <tbody>
          <tr>
            <XUIEditableTableCellAutocompleter {...spreadProps} />
          </tr>
        </tbody>
      </table>,
    );
    expect(wrapper.find(XUIAutocompleter).props()).toMatchObject(spreadProps);
  });

  it('fires the onOpen and onClose handlers when the top-level methods are fired', () => {
    const onOpenMock = jest.fn();
    const onCloseMock = jest.fn();
    const wrapper = mount(
      <table>
        <tbody>
          <tr>
            <XUIEditableTableCellAutocompleter
              onClose={onCloseMock}
              onOpen={onOpenMock}
              onSearch={() => {}}
            />
          </tr>
        </tbody>
      </table>,
    );

    const cellAuto = wrapper.find(XUIEditableTableCellAutocompleter);
    cellAuto.instance().openDropdown();
    expect(onOpenMock).toHaveBeenCalledTimes(1);

    cellAuto.instance().closeDropdown();
    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });
});
