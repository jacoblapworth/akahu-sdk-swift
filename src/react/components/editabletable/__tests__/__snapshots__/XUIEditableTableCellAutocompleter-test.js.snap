// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<XUIEditableTableCellAutocompleter /> renders correctly 1`] = `
<XUIEditableTableCellControl
  cellIds={
    Object {
      "control": undefined,
      "wrapper": undefined,
    }
  }
  className="xui-editabletablecellautocompleter"
  isFocused={false}
  onClick={[Function]}
>
  <XUIAutocompleter
    className="xui-editabletablecellautocompleter--control"
    closeOnTab={true}
    disableWrapPills={false}
    dropdownFixedWidth={false}
    forceDesktop={true}
    inputProps={
      Object {
        "onBlur": [Function],
        "onFocus": [Function],
      }
    }
    isInputLabelHidden={true}
    isLegacyDisplay={true}
    loading={false}
    matchTriggerWidth={true}
    onSearch={[Function]}
    openOnFocus={false}
    searchDebounceTimeout={200}
    triggerClassName="xui-editabletablecellautocompleter--trigger"
  />
</XUIEditableTableCellControl>
`;
